design-proposals/vm-rollout-strategy/vm-rollout-strategy.md

```diff
@@ -0,0 +1,205 @@
+# Overview
+
+Today changes to a VM object (i.e. changing VM.spec.template) get applied, but only take effect upon the next restart - we call these pending changes “staged”. This is a proposal to introduce a custer-level rollout strategy field to control how changes to the VM object will propagate to the VirtualMachineInstance object without interupting the workload.
+
+
+## Motivation
+
+In most cases, the user would prefer the changes to the VM object to eventually propagate to the VMI object, as long as the rollout does not interrupt the workload. This behavior would be similar to how the Deployments roll out ReplicaSets.
+
+
+## Goals
```

Comment:
How is this going to work/sync with the network hotplug interfaces?
At the moment, it has a FG and is not controlled by any strategy (it is by default `LiveUpdate`).

I am in favor to graduate the network hotplug in v1.2 or v1.3. depending on user feedback.
But if we bind it to this new FG or to the rollout strategy parameter, we may have a backward compatibility issue on our hand.

Reply:
Yes, I understand the potential complications.

I think the network hotplug interface will need to preserve its current enablement.
We definitely need to have a follow-up discussion about the harmonization of all hot pluggable interfaces (Network and Storage) - how can we make all interfaces be driven by the rollout strategy.

The fact that the the network hotplug is behind a feature gate means that we didn't commit to this API and effectively don't have to provide an upgrade path if we are unable to find one.

Reply:
We need to align all hotplug mechanisms.

@mhenriks is currently working to prepare storage hotplug for a unified UX.

Before you @EdDev can GA network, we should sync and see how we can provide this unified UX.

IOW, you want to hold you GA until it's clear how a unified UX looks.

Reply:
I understand the direction, but the generalization of this proposal and the actual progress of things are somehow colliding.
- The proposal attempts to generalize the VM rollout strategy, using only two examples: CPU & Memory.
   It does not mention storage or networking, nor a plan on how to consolidate them into the same strategy.
- Networking hotplug is not formally GA, but it will be very difficult to break it at the moment based on its (perhaps premature) adoption. I think this needs to be further discussed to detect a possible solution in the future.

IMO, if there is a time to plan for the harmonization, it is now. This proposal de-facto defines the minimum needed for it to occur.

Reply:
@EdDev Hi, I think that generalization of this proposal should be in a dedicated design proposal, something like "dynamic VM configurtation". In that proposal we can define graduation phases, and connect the subjects like hotplug harmonization to these phases.

Reply:
@EdDev there are two main points in this proposal

1. be declarative and permit eventual consistency, incl defining an error condition
2. have a minimal api

To me this is not mem or cpu specific, and I'm also not - at a first glance - seeing how this collides with network.

Ain't these two points for a GA'ed network hotplug desirable as well?

Reply:
> Ain't these two points for a GA'ed network hotplug desirable as well?

The goals are, the solution was more aggressive with networking.
There we are opinionated and avoid the staging option.
At the time, we thought it makes sense. I still think it makes sense now, but it is hard to argue with a desire to have minimal control as suggested in this proposal with the strategy. What I am saying is that the defaults collide.

Reply:
Ok. it would be great if network can align to this, and then we are good.